packages/router/history.ts

```diff
@@ -570,7 +570,7 @@ function getUrlBasedHistory(
     }
 
     if (v5Compat && listener) {
-      listener({ action, location });
+      listener({ action, location: history.location });
```

Comment:
History should pass the encoded location (from `window.location` after `pushState`) and not the incoming unencoded location